std/algorithm/sorting.d

```diff
@@ -1225,26 +1221,37 @@ private template TimSortImpl(alias pred, R)
             stack[stackLen++] = Slice(i, runLen);
             i += runLen;
 
-            // Collapse stack so that (e1 >= e2 + e3 && e2 >= e3)
+            // Collapse stack so that (e1 > e2 + e3 && e2 > e3)
             // STACK is | ... e1 e2 e3 >
             while (stackLen > 1)
             {
-                immutable run3 = stackLen - 1;
-                immutable run2 = stackLen - 2;
-                immutable run1 = stackLen - 3;
-                if (stackLen >= 3 && stack[run1].length <= stack[run2].length + stack[run3].length)
+                immutable run1 = stackLen - 4;
+                immutable run2 = stackLen - 3;
+                immutable run3 = stackLen - 2;
+                immutable run4 = stackLen - 1;
+                
+                if (stackLen > 2 && stack[run2].length <= stack[run3].length + stack[run4].length || 
+                    stackLen > 3 && stack[run1].length <= stack[run3].length + stack[run2].length )
                 {
-                    immutable at = stack[run1].length <= stack[run3].length
-                        ? run1 : run2;
+                    immutable at = stack[run2].length < stack[run4].length ? run2 : run3;
                     mergeAt(range, stack[0 .. stackLen], at, minGallop, temp);
-                    --stackLen;
+                    
                 }
-                else if (stack[run2].length <= stack[run3].length)
+                else if (stack[run3].length > stack[run4].length) break;
+                else mergeAt(range, stack[0 .. stackLen], run3, minGallop, temp);
+                
+                stackLen -= 1;
+            }
+            
+            // Assert that the code above established the invariant correctly
+            debug
```

Comment:
Perhaps extract it into an inner nested function? Not sure if that would have any performance impact once inlining is enabled though (it would allow you to add the contract there though).